### ISSUE TITLE
Baresip with Coturn relay not working?

### ISSUE THREAD

**Issue by Syirrus** [alfredh/baresip#979]
I'm not sure if this is an issue with Baresip or my noob-ness, but I have setup Coturn so that I can use STUN and TURN with Baresip. I have tested my Coturn server with Trickle ICE and it appears to be relaying (TURN) and resolving host IP's (STUN) properly.  

Here is my `accounts` file:

_"Pi" <sip:101@domain.com>;auth_pass=pass1234;stunuser=guest;stunpass=somepassword;stunserver=turn:domain.com:3478;medianat=turn_

My understanding is that if I specify turn as a stun server, I will only have relay capabilities even though Coturn does both STUN and TURN. Is that correct?  

If I use these settings baresip loads as follows:

_using stunserver: 'turn:domain.com:3478'
101@domain.com: Using medianat 'turn'
.....
[h264_omx @ 0xac200470] Using OMX.broadcom.video_encode
stream: incoming rtp for 'audio' established, receiving from domain.com:10842
sip:102@domain.com: session closed: Connection reset by peer0_

However, if I try to make a video call over two different networks 1 NAT and the other over a cellular network (I assume that is NAT'ed) I only get any video, which if TURN is working, it should be relaying audio and video. If I'm on the same NAT'ed network, I get video. 

Do both clients need to support TURN and STUN?  I tried with microsip and baresip. 

Again, the expected behavior is to have both STUN and TURN working together, so that if STUN fails, TURN is there as a backup to ensure both audio and video make it though.

**Comment by Syirrus** [comment]
@alfredh 
Okay, for the past two days I have been testing and testing. I believe I have narrowed the problem down slightly. If I use two different SIP clients, MicroSIP and GS Wave for Android, video call work consistently, even on differing networks. When I add Baresip into the equation with either MicroSIP or GS wave, video works intermittently. 

Here is the layout of my network which is NOT effecting the intermittent video problem.

PBXSRV ----------- NAT --- BareSIP 
                                                    |
                                                 MicroSIP

This is an ideal setup and video calls work intermittently.  If I replace Baresip with GS Wave it works every time. This is without using a Coturn or a google STUN server. 

Any ideas?

Syirrus

**Comment by sreimers** [comment]
> My understanding is that if I specify turn as a stun server, I will only have relay capabilities even though Coturn does both STUN and TURN. Is that correct?

`medianat=turn` means TURN only.

> Again, the expected behavior is to have both STUN and TURN working together, so that if STUN fails, TURN is there as a backup to ensure both audio and video make it though.

For this setup you need to configure `medianat=ice`

**Comment by Syirrus** [comment]
> > My understanding is that if I specify turn as a stun server, I will only have relay capabilities even though Coturn does both STUN and TURN. Is that correct?
> 
> `medianat=turn` means TURN only.
> 
> > Again, the expected behavior is to have both STUN and TURN working together, so that if STUN fails, TURN is there as a backup to ensure both audio and video make it though.
> 
> For this setup you need to configure `medianat=ice` , but the peer needs ICE support too.
@sreimers 
Thank you for your response.  I have tried with medianat=ice and when I do so I don't get video at all, even though my TURN server appears to be setup properly according to Trickle ICE.  I wanted to force TURN to see if my video problems were related to NAT traversal issues.  They are not and seem to be linked to baresip, because if I use differing SIP clients without Baresip (even without TURN), I get video consistently.  The funny thing is, when I accept or initiate a video call on the Pi zero, the led on the camera comes on.  I just don't see the video on the other end :(

**Comment by Syirrus** [comment]
Well after much fighting, I'm stuck.  Video & Audio work on the same LAN that is NAT'ed (public IP facing PBX). If I switch to the cellular network the calls connect and I get audio, but no video. I've used UDP, and tried TLS to no avail.

**Comment by Syirrus** [comment]
Well, I believe I can safely conclude that this issue lies within Baresip :(. If I use different SIP clients other than Baresip (Microsip and GS wave for android) video calls work over my (NAT'ed) network and the cellular network with GS Wave for android. Those clients also work both on my NAT'ed network as well.

As a part of the testing, I removed the STUN and TURN server out of the equation completely and have everything going over UDP.

When to try with Baresip on my (NAT'ed) network with GS Wave or Linphone over the same (NAT'ed) network, video calling works most of the time, but it's intermittent (irrespective of which device is calling whom).  My camera on the pi indicates that it's on always, but sometimes there is just not picture on GS Wave or Linphone. I would say it works 2 out of 3 times.

When I try with Baresip on my (NAT'ed) network with GS Wave over the cellular network, video calling fails every time, however I do have audio, just not video.

Any ideas @alfredh ?  I have spent tons of time on this double and triple checking my config and accounts file and it's just not that.
